### ISSUE TITLE
Allow the `object` or `property` option to be omitted with `no-restricted-properties`

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#7137]
**What rule do you want to change?**

[`no-restricted-properties`](http://eslint.org/docs/rules/no-restricted-properties)

**Does this change cause the rule to produce more or fewer warnings?**

The default behavior of the rule would be unchanged.

**How will the change be implemented? (New option, new default behavior, etc.)? What will the rule do after it's changed?**

Currently, `no-restricted-properties` accepts a list of objects with the `object` and `property` keys. Both options are required.

It would be useful to be able to disallow access to a property on *every* object, rather than having to provide explicit object names. For example, one might want to disallow use of the deprecated `__defineGetter__` property, which is currently not possible with ESLint.

Similarly, it would be useful to be able to disallow all property access for a particular object, without specifying property names. For example, one might want to encourage use of the `require()` function, and disallow access to `require.resolve()`, `require.cache`, etc.

I propose that we make the `object` and `property` arguments optional. If `property` is provided and `object` is omitted, an error is reported whenever `property` is accessed from any object. If `object` is provided and `property` is omitted, an error is reported whenever any property of `object` is accessed.

If both `object` and `property` are omitted, ESLint should raise an error about an invalid schema. (The consistent behavior would be to disallow all forms of property access, but it's unlikely that anyone would actually want to do this and so it would cause confusion.)

**Please provide some example code that this change will affect:**

```js
/* eslint no-restricted-properties: [2, {"property": "__defineGetter__"}] */

foo.__defineGetter__(bar, baz); // report error
bar.__defineGetter__(bar, baz); // report error
```

```js
/* eslint no-restricted-properties: [2, {"object": "require"}] */

require('foo') // ok
require.resolve('foo') // report error
require.cache // report error
require['foo'] // report error
```

**What does the rule currently do for this code?**

It causes ESLint to throw an error because the schema is considered invalid.

**Comment by ljharb** [comment]
My use case is to block `__proto__`, all the legacy getter/setter methods, and to block `splice`, all on any object.

**Comment by not-an-aardvark** [comment]
For what it's worth, you can already block `__proto__` with [`no-proto`](http://eslint.org/docs/rules/no-proto).

**Comment by ljharb** [comment]
With this change, `no-proto` could be deprecated in favor of the more generic rule :-p or at least reimplemented in terms of it.

**Comment by not-an-aardvark** [comment]
Decided to go ahead and implement this; see #7139.

**Comment by vitorbal** [comment]
Just a thought: could we use `"*"` or something similar when we want to signify "any object" or "any property"?
```yaml
no-restricted-properties: [2, {"object": "*", "property": "__defineGetter__"}]
```

Another thought, what if `object` and `property` could take a regular expression? Would that be useful?

**Comment by not-an-aardvark** [comment]
@vitorbal The issue with wildcard strings is that they're indistinguishable from properties. For example, the following code currently reports an error:

```js
/* eslint no-restricted-properties: [2, {"object": "foo", "property": "*"}] */

foo['*']

```

**Comment by platinumazure** [comment]
I'd prefer either omitting the property or using `null` as a wildcard (well, as an "unspecified" token I guess?). Just my two cents.

**Comment by vitorbal** [comment]
@not-an-aardvark ah, makes sense, thanks for the explanation! I don't have a strong opinion about this so I would be fine with omitting the property as well.

**Comment by alecharmon** [comment]
Wouldn't it be even clearer to introduce an  'all-objects' option that takes a bool rather than leave it to the ambiguity of null?

**Comment by ljharb** [comment]
@alecharmon and also `all-properties`? The current proposal seems the clearest to me.

**Comment by platinumazure** [comment]
Ah, I didn't realize JSON Schema would let us require just one or the other (in conjunction with anyOf). Okay, I like the original proposal (just omit the config properties as needed) better and withdraw my suggestion of null valuing.
